File: solution-1/server/event_source.py
from asyncio import Queue
from collections import defaultdict
from typing import DefaultDict

class EventSource:
    def __init__(self):
        self.mapStreams: DefaultDict[str, Queue] = defaultdict()

    def newStream(self, channel_id: str):
        self.mapStreams[channel_id] = Queue()

    async def add(self, channel_id: str, data: str):
        if channel_id not in self.mapStreams:
            print(f"Channel {channel_id} not found")
            return
        streamer = self.mapStreams[channel_id]
        await streamer.put(data)

    async def listen(self, channel_id: str):
        if channel_id not in self.mapStreams:
            print(f"Channel {channel_id} not found")
            return
        streamer = self.mapStreams[channel_id]
        while True:
            val = streamer.get()
            yield val
            if val == '<end>':
                self.mapStreams.pop(channel_id)
                return

File: playground/eventSource.ts
import type { Stream } from 'node:stream';

// Usage:
//   eventSource.newStream(channelId, res);
//
//   eventSource.listen(channelId, (data: any) => {
//      /* do something with the data */
//   });
//
//    Somewhere else in the code:
//    for {
//      eventSource.send(id, response.data);
//    }

export class EventSource {
  private mapStreams: Map<string, Stream>;

  constructor() {
    this.mapStreams = new Map();
  }

  newStream(channelId: string, streamer: Stream) {
    this.mapStreams.set(channelId, streamer);
  }

  send(channelId: string, data: any) {
    if (!this.mapStreams.has(channelId)) {
      console.log(`Channel ${channelId} not found`);
      return;
    }
    const streamer = this.mapStreams.get(channelId);
    streamer?.emit('message', data);
  }

  listen(channelId: string, callback: (data: string) => void) {
    const streamer = this.mapStreams.get(channelId) as Stream;
    if (!streamer) {
      console.log(`Channel ${channelId} not found`);
      return;
    }

    streamer.on('message', (data) => {
      callback(data);
    });
  }

  close(channelId: string) {
    const stream = this.mapStreams.get(channelId);
    if (!stream) {
      console.log(`Channel ${channelId} not found`);
      return;
    }
    stream.removeAllListeners();
    this.mapStreams.delete(channelId);
  }
}
